src/core/compiler.jl

```diff
@@ -288,49 +229,144 @@ function build_model_info(input_expr)
 end
 
 """
-    translate_tilde!(model_info)
+    replace_vi!(model_info)
+
+Replaces @varinfo() expressions to the VarInfo instance.
+"""
+function replace_vi!(model_info)
+    ex = model_info[:main_body]
+    vi = model_info[:main_body_names][:vi]
+    ex = MacroTools.postwalk(x -> @capture(x, @varinfo()) ? vi : x, ex)
+    model_info[:main_body] = ex
+    return model_info
+end
+
+"""
+    replace_logpdf!(model_info)
+
+Replaces @logpdf() expressions to the VarInfo instance.
+"""
+function replace_logpdf!(model_info)
+    ex = model_info[:main_body]
+    vi = model_info[:main_body_names][:vi]
+    ex = MacroTools.postwalk(x -> @capture(x, @logpdf()) ? :($vi.logp) : x, ex)
+    model_info[:main_body] = ex
+    return model_info
+end
+
+"""
+    replace_sampler!(model_info)
 
-Translates ~ expressions to observation or assumption expressions, updating `model_info`.
+Replaces @sampler() expressions to the VarInfo instance.
 """
-function translate_tilde!(model_info)
+function replace_sampler!(model_info)
     ex = model_info[:main_body]
+    spl = model_info[:main_body_names][:sampler]
+    ex = MacroTools.postwalk(x -> @capture(x, @sampler()) ? spl : x, ex)
+    model_info[:main_body] = ex
+    return model_info
+end
+
+"""
+\"""
+    replace_tilde!(model_info)
+
+Replaces ~ expressions with observation or assumption expressions, updating `model_info`.
+\"""
+function replace_tilde!(model_info)
+    ex = model_info[:main_body]
+    ex = MacroTools.postwalk(x -> @capture(x, @. L_ ~ R_) ? dot_tilde(L, R, model_info) : x, ex)
+    $(VERSION >= v"1.1" ? "ex = MacroTools.postwalk(x -> @capture(x, L_ .~ R_) ? dot_tilde(L, R, model_info) : x, ex)" : "")
     ex = MacroTools.postwalk(x -> @capture(x, L_ ~ R_) ? tilde(L, R, model_info) : x, ex)
     model_info[:main_body] = ex
     return model_info
 end
+""" |> Meta.parse |> eval
 
 """
-    update_args!(model_info)
+    tilde(left, right, model_info)
 
-Extracts default argument values and replaces them with `nothing`.
+The `tilde` function generates observation expression for data variables and assumption expressions for parameter variables, updating `model_info` in the process.
```

Comment:
90 chars. Also, maybe consider replacing "observation" with `observe` and `assumption with `assume` to more tightly reflect internal naming scheme.